energy-integration/mex-governance/src/lib.rs

```diff
@@ -0,0 +1,202 @@
+        let voting_week = current_week + 1;
+
+        self.advance_week_if_needed(voting_week);
+
+        let user_votes_in_week_mapper = self.user_votes_in_week(user_id, voting_week);
+        require!(
+            user_votes_in_week_mapper.is_empty(),
+            ALREADY_VOTED_THIS_WEEK
+        );
+
+        let user_energy = self.get_energy_amount_non_zero(&caller);
+
+        let mut total_vote_amount = BigUint::zero();
+        let mut farm_votes = ManagedVec::new();
+        let mut farm_votes_event = ManagedVec::new();
+        for vote in votes {
+            let (farm_address, amount) = vote.into_tuple();
+            let farm_id = self.farm_ids().get_id_non_zero(&farm_address);
+
+            require!(
+                !self.blacklisted_farms().contains(&farm_id),
+                FARM_BLACKLISTED
+            );
+            require!(
+                self.whitelisted_farms().contains(&farm_id),
+                FARM_NOT_WHITELISTED
+            );
+
+            self.farm_votes_for_week(farm_id, voting_week)
+                .update(|sum| *sum += amount.clone());
+            self.voted_farms_for_week(voting_week).insert(farm_id);
+
+            total_vote_amount += &amount;
+
+            farm_votes.push(FarmVote {
+                farm_id,
+                vote_amount: amount.clone(),
+            });
+            farm_votes_event.push(FarmVoteView {
+                farm_address,
+                farm_emission: amount,
+            });
+        }
+
+        require!(total_vote_amount == user_energy, INVALID_VOTE_AMOUNT);
+
+        self.total_energy_voted(voting_week)
+            .update(|sum| *sum += &user_energy);
+        user_votes_in_week_mapper.set(&farm_votes);
+
+        self.emit_vote_event(voting_week, farm_votes_event);
+    }
+
+    #[payable("*")]
+    #[endpoint(incentivizeFarm)]
+    fn incentivize_farm(
+        &self,
+        farms_incentives: MultiValueEncoded<MultiValue3<ManagedAddress, BigUint, Week>>,
+    ) {
+        let current_week = self.get_current_week();
+        let mut remaining_payment = self.call_value().single_esdt();
+        let incentive_payment_token = self.incentive_token().get();
+        require!(
+            remaining_payment.token_identifier == incentive_payment_token,
+            INVALID_INCENTIVE_PAYMENT
+        );
+
+        for farm_incentive in farms_incentives {
+            let (farm_address, farm_incentive, week) = farm_incentive.into_tuple();
+            require!(week > current_week, INVALID_INCENTIVE_WEEK);
+            require!(
+                remaining_payment.amount >= farm_incentive,
+                INVALID_INCENTIVE_PAYMENT
+            );
+            remaining_payment.amount -= &farm_incentive;
+
+            let farm_id = self.farm_ids().get_id_non_zero(&farm_address);
+
+            self.farm_incentive_for_week(farm_id, week)
+                .update(|sum| *sum += farm_incentive);
+        }
+
+        if remaining_payment.amount > BigUint::zero() {
```

Comment:
The framework does this for you anyway. Code is more readable this way.

```suggestion
        if remaining_payment.amount > 0 {
```

Reply:
Done.